nautobot_golden_config/utilities/helper.py

```diff
@@ -139,3 +150,103 @@ def get_device_to_settings_map(queryset):
+        module_name, model_name = obj_type.split(".")
+    except ValueError:  # not enough values to unpack (expected 2, got 1). We get this error if there is no "." in the obj_type
+        return f"Incorrect object type format {obj_type}. The correct format is 'module_name.model_name'."
+    try:
+        model = apps.get_model(module_name, model_name)
+    except ObjectDoesNotExist:
+        return f"Model {model_name} does not exist."
+    # get the object behind the model using the pk
+    try:
+        obj = model.objects.get(id=obj_id)
+    except ObjectDoesNotExist:
+        return f"Object wit ID {obj_id} does not exist."
+
+    # Get user permissions, terminate early if they do not have permission to view
+    app_label = obj._meta.app_label
+    model_name = obj._meta.model_name
+    permission_required = f"{app_label}.view_{model_name}"
+    # Bypass restriction for superusers and exempt views
+    if user.is_superuser or permission_is_exempt(permission_required):
+        pass
+    # User is anonymous or has not been granted the requisite permission
+    elif not user.is_authenticated or permission_required not in user.get_all_permissions():
+        return "You have no permission to read secrets. This incident will be reported."
+
+    # look through the attributes of the model to find if it has an id relationship
+    secrets_group = None
+    secrets_group_id = obj.__dict__.get("secrets_group_id")
+    if secrets_group_id:
+        secrets_group = SecretsGroup.objects.get(id=secrets_group_id)
+
+    # if the secret is not in the attributes, it may be in the relationships
+    for value in obj.get_relationships_data()["destination"].values():
+        # Find the relationship to secrets group
+        if isinstance(value["value"], SecretsGroup):
+            secrets_group = SecretsGroup.objects.get(slug=value["value"])
+            break
+
+    if secrets_group:
+        return secrets_group.get_secret_value(
+            access_type=secret_access_type,
+            secret_type=secret_type,
+            obj=obj,
+        )
+
+    return None
+
+
+def render_secrets(config_to_render: str, user: User):
+    """Renders secrets using the get_secrets filter, and Netutils ones."""
+    jinja_env = jinja2.Environment(autoescape=True)
+
+    # Wrapper for get_secret filter that includes user argument to ensure
+    # that secrets are only rendered by authorized users.
+    user_gets_secret = partial(get_secret, user)
+
+    for name, func in jinja2_convenience_function().items():
+        jinja_env.filters[name] = func
+
+    jinja_env.filters["get_secret"] = user_gets_secret
+    template = jinja_env.from_string(config_to_render)
+    return template.render()
+
+
+def get_config_to_push(config_to_render: str, user: User, device: Device):
+    """Renders final configuration push artifact from intended configuration."""
+    config_to_push = render_secrets(config_to_render, user)
+
+    if PLUGIN_CFG.get("config_push_processing"):
```

Comment:
I was thinking of a pattern like
```
# Have a default list of callable in a specific order
CONFIG_PUSH_CALLABLE = [render_secrets, missing_only]
if PLUGIN_CFG.get("config_push_callable"):
    CONFIG_PUSH_CALLABLE =PLUGIN_CFG["config_push_callable"]

# Subscribe to which ones you use
CONFIG_PUSH_SUBSCRIBED = [render_secrets]
if PLUGIN_CFG.get("config_push_subscribed"):
    CONFIG_PUSH_SUBSCRIBED =PLUGIN_CFG["config_push_subscribed"]


def get_config_to_push(details_not_shown):
    for callable in CONFIG_PUSH_CALLABLE:
        if CONFIG_PUSH_SUBSCRIBED.get(callable.__str__):
            callable(some_signature)

```

This way, each person can subscribe to what they want to, and we can have sane defaults.

Reply:
I see it a bit overcomplicated... I think the main difference is that I was approaching it only as a "full" config management, and you are adding the "per feature" config.
Could this "per feature" function be done on a different method, using the "full" one as a reference? This would simply a lot this function, but it would require another one implementing your suggested approach.
Open to do one way or another.

Reply:
These are the use cases I am seeing as valid.
https://github.com/nautobot/nautobot-plugin-golden-config/issues/255#issuecomment-1225717861

Complication is going to exists, we have to accept that at this point. It's understanding where to put the complication. I did not consider doing a separate full vs partial, and not sure I want that... just yet.

Reply:
Ok, if we want to chain all these methods, they should share the same signature. They could even get not only the intended configuration, but ALL the configs (such as backup), to do some remediation processing.
However, something they should pass is the "config_to_push". So, initially the "config_to_push" could be empty, and every method would take it as the starting point, to finally return the processed one. 
This is the approach I'm going to take, so the signature of the callables would be:
`render_secrets(configs, user: User, device: Device, reference_config: str)`
So, if the callable is the first one in the chain, the `reference_config` would be `""`,  and it will use the `configs.intended_configuration` as the intended one. If not, it should use this `reference_config` as its intended reference configuration.
Do you see any flaw point in the proposal?

Reply:
ok, I think the delta to what you are suggesting vs my initial is:
* configs is a complex object, that will include `configs.intended_configuration` and similar
* config_to_push is what is returned

That makes sense, I like the idea of `config_to_push`, I just don't see if in the signature, in theory you need to send it no matter what??

Reply:
I implemented it with the 2 steps (define + subscribe), but I still think that with only one step defining both it would be enough (taking into account the use cases you mentioned)

Reply:
I got a much simpler signature that I expected:
`my_callable_function(config_to_push: str, configs: models.GoldenConfig, request: HttpRequest)`

Reply:
Yes, that seems to make sense to me right now :) . Love seeing this being fleshed out!!